### PR TITLE
[BEAM-10961] Make plugins' transitive dependencies compileOnly dependencies.

### PR DESCRIPTION
**Pull request by sonam-vend**
**Please** add a meaningful description for your change here

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Dataflow/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by sonam-vend** [comment]
Run PythonDocs PreCommit

**Comment by sonam-vend** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by sonam-vend** [comment]
Run PythonDocs PreCommit

**Comment by sonam-vend** [comment]
Run Python PreCommit

**Comment by sonam-vend** [comment]
Run Python PreCommit

**Comment by sonam-vend** [comment]
Run Python PreCommit

**Comment by sonam-vend** [comment]
Run PythonDocs PreCommit

**Review by ibzib — APPROVED** [review]
LGTM

**Comment by ibzib** [comment]
R: @kennknowles

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 1]
```diff
@@ -855,9 +855,11 @@ class BeamModulePlugin implements Plugin<Project> {
       // configurations because they are never required to be shaded or become a
       // dependency of the output.
       def compileOnlyAnnotationDeps = [
-        "com.google.auto.value:auto-value-annotations:1.7",
+        "com.google.auto.value:auto-value-annotations:1.7.2",
         "com.google.auto.service:auto-service-annotations:1.0-rc6",
         "com.google.j2objc:j2objc-annotations:1.3",
+        "org.checkerframework:checker-qual:3.7.0",
```
Comment (kennknowles) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:861]: These annotations do end up in the class files so users can benefit from them.

Java spec is that annotations that cannot be resolved on the classpath should be ignored. Unfortunately, the JDK has a bug (don't recall if/when it was fixed) where it actually causes a crash.

See https://github.com/apache/beam/pull/13169 and https://issues.apache.org/jira/browse/BEAM-10632

Comment (kennknowles) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:861]: If JDK worked properly, then users who had check qual on their classpath would get the benefit, while other users would not be impacted.

Comment (kennknowles) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:861]: So instead, I think the checker-qual needs to be `compile` scope and right here add it to the list of `allowUnusedDeclared`.

Comment (sonam-vend) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:861]: @kennknowles If I understood your point correctly, you mean to remove checker-qual from compileOnlyAnnotationDeps,

and add here like,
 ```
// Disregard unused but declared (test) compile only dependencies used
      // for common annotation classes used during compilation such as annotation
      // processing or post validation such as spotbugs.
      project.dependencies {
        compileOnlyAnnotationDeps.each { dep ->
          permitUnusedDeclared dep
          permitTestUnusedDeclared dep
        }
        **permitUnusedDeclared "org.checkerframework:checker-qual:$checkerframework_version"**
      }
```

Comment (kennknowles) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:861]: Yes, and also add it to `compile` scope deps.

Comment (ibzib) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:861]: Isn't it already in `compile`? https://github.com/apache/beam/blob/7aa7e13c564674a99c99a08b21c97ce92111652a/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy#L911

Comment (sonam-vend) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:861]: @kennknowles , But it is already defined in compile scope https://github.com/apache/beam/blob/master/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy#L911

Comment (kennknowles) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:861]: Yea that's what I thought. OK then that part should be fine.

Comment (sonam-vend) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: corrected as your suggestion @kennknowles

**Comment by sonam-vend** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by sonam-vend** [comment]
Run Java_Examples_Dataflow PreCommit

**Comment by sonam-vend** [comment]
Run Python PreCommit

**Comment by sonam-vend** [comment]
Run Java PreCommit

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 2]
```diff
@@ -855,9 +855,10 @@ class BeamModulePlugin implements Plugin<Project> {
       // configurations because they are never required to be shaded or become a
       // dependency of the output.
       def compileOnlyAnnotationDeps = [
-        "com.google.auto.value:auto-value-annotations:1.7",
+        "com.google.auto.value:auto-value-annotations:1.7.2",
```
Comment (ibzib) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:858]: It looks like there are a couple places that use `com.google.auto.value:auto-value:1.7`, can you update that to 1.7.2 as well? And all of these places should probably use an `autovalue_version` variable to make sure they are using the same version.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 3]
```diff
@@ -855,9 +855,10 @@ class BeamModulePlugin implements Plugin<Project> {
...
         "com.google.auto.service:auto-service-annotations:1.0-rc6",
         "com.google.j2objc:j2objc-annotations:1.3",
+        "com.google.code.findbugs:jsr305:3.0.2",
```
Comment (ibzib) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:861]: Can you create a jsr305_version variable to share with the jsr305 library defined in `project.ext.library`?

Comment (sonam-vend) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Done.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 2 above)
Comment (sonam-vend) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: done.

**Comment by sonam-vend** [comment]
Run Java_Examples_Dataflow PreCommit

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 4]
```diff
@@ -452,6 +452,8 @@ class BeamModulePlugin implements Plugin<Project> {
     def spark_version = "2.4.7"
     def spotbugs_version = "4.0.6"
     def testcontainers_version = "1.15.1"
+    def jsr305_version = "3.0.2"
```
Comment (ibzib) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:455]: Nit: can you insert the new lines according to alphabetical order?

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 5]
```diff
@@ -858,9 +860,10 @@ class BeamModulePlugin implements Plugin<Project> {
       // configurations because they are never required to be shaded or become a
       // dependency of the output.
       def compileOnlyAnnotationDeps = [
-        "com.google.auto.value:auto-value-annotations:1.7",
+        "com.google.auto.value:auto-value-annotations:$autovalue_version",
```
Comment (ibzib) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:863]: Nit: can you insert the new lines according to alphabetical order?

**Review by ibzib — COMMENTED** [review]
Thanks, this looks great. One general comment: since the order of dependency lists doesn't matter to the computer, please keep them alphabetized where possible. It makes them easier to read for humans. (This also applies to all of the other PRs in this series.)

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 5 above)
Comment (ibzib) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:863]: (I realize this one wasn't in exact alphabetical order to begin with, but it should have been.)

**Pull request merged**